src/traces/surface/convert.js

```diff
@@ -349,19 +352,54 @@ proto.refineCoords = function(coords) {
     }
 };
 
+function insertIfNewLevel(arr, newValue) {
+    var found = false;
+    for(var k = 0; k < arr.length; k++) {
+        if(newValue === arr[k]) {
+            found = true;
+            break;
+        }
+    }
+    if(found === false) arr.push(newValue);
+}
+
 proto.setContourLevels = function() {
-    var nlevels = [[], [], []];
+    var newLevels = [[], [], []];
+    var useNewLevels = [false, false, false];
     var needsUpdate = false;
 
-    for(var i = 0; i < 3; ++i) {
+    var i, j, value;
+
+    for(i = 0; i < 3; ++i) {
         if(this.showContour[i]) {
             needsUpdate = true;
-            nlevels[i] = this.scene.contourLevels[i];
+
+            if(
+                this.contourSize[i] > 0 &&
+                this.contourStart[i] !== null &&
+                this.contourEnd[i] !== null &&
+                this.contourEnd[i] > this.contourStart[i]
+            ) {
+                useNewLevels[i] = true;
+
+                for(j = this.contourStart[i]; j < this.contourEnd[i]; j += this.contourSize[i]) {
+                    value = j * this.scene.dataScale[i];
```

Comment:
Have you tried re-using [`setContours`](https://github.com/plotly/plotly.js/blob/master/src/traces/contour/set_contours.js#L16)? It probably gives similar results, but it would be nice to double-check.

Reply:
I tried that on a different branch. Also wanted to add `autocontours` and `ncontours` in this PR. 
But unfortunately that could add too much complexity while surface contours are in 3 directions not only z.